### ISSUE TITLE
Use the standard enrollment table instead of a custom CCX table

### ISSUE THREAD

**Issue by pdpinch** [jazkarta/edx-platform#78]
Depends on #43

**Comment by cewing** [comment]
Now that the custom opaque-key implementation for ccx is in place, this should be approachable.

**Comment by cguardia** [comment]
Ok, I will be taking care of this issue. I'm not familiar with it, so first I'll review the code and see what's required.

**Comment by pdpinch** [comment]
Dave Ormsbee keeps claiming that all we need to do is remove the code for the custom CCX enrollment table.

**Comment by cguardia** [comment]
Started work on this. After removing membership models, there are a bunch of tests that need fixing, so I'm doing that now.

**Comment by cguardia** [comment]
In addition to the tests, I've been having to rewire functionality from the enrollment pages that required queries to the ccx membership tables. So it was not just a matter of removing the models, because they power up the whole membership tab and some other places, like the list of currently subscribed ccx courses. I have not finished test fixing either, so I will be working on this for at least the rest of the week.

**Comment by bdero** [comment]
@cguardia Let me know when this is ready for @pwilkins and I to test on a sandbox and we'll give it a spin.

**Comment by cguardia** [comment]
I'm running a bit behind, but shooting for Thursday morning.

**Comment by cguardia** [comment]
I still have more work to do here, plus #80. Will keep you posted.

**Comment by cguardia** [comment]
Sorry that this has taken me more time than expected. I had a meeting with @cewing late last week and that helped me get unstuck. I have pushed my progress to this branch:

https://github.com/jazkarta/edx-platform/tree/remove-ccx-enrollment

There is only one failing ccx test at this point. I will keep working on this today, but I think it can be tested now.

**Comment by cguardia** [comment]
All tests are passing now. This is definitely ready for testing. I'll work on #80 while you try it out.

**Comment by pdpinch** [comment]
Thanks Carlos. We are reviewing this now on sandbox2o.mitx.mit.edu

**Comment by pdpinch** [comment]
Carlos, we're getting an error on creating CCXs -- although we're running off yestersday's version of your branch. Do you think a pull and re-deploy will fix this?

https://gist.github.com/bdero/d63c3c0cc393bd26e9b0

**Comment by cewing** [comment]
Looks to me like a typo in the `create_course` view:

https://gist.github.com/bdero/d63c3c0cc393bd26e9b0#file-ccx_traceback-L11

The comma in the `kwargs` argument should be a colon.

**Comment by cguardia** [comment]
Yes, @cewing is right. Fixed the typo. The reason I didn't get that on my testing is that this url is only used when trying to create a ccx from a course with a deprected id, which is not allowed.

**Comment by pdpinch** [comment]
With a completely new student, I'm getting duplicate enrollments in my course dashboard -- although they both lead to the same URL (https://sandbox2o.mitx.mit.edu/courses/ccx-v1:ODLEng+DemoX+2015_T2+ccx@37/info) ??

<img width="879" alt="screen shot 2015-07-14 at 8 25 04 pm" src="https://cloud.githubusercontent.com/assets/430126/8688166/9294e1c0-2a66-11e5-8a17-5f3e02581b1e.png">

Is this because we are double-enrolling the student -- in the root course and in the CCX? Is that still necessary?

**Comment by cguardia** [comment]
I think we could probably remove our custom list from the dashboard. What do you think @cewing?

**Comment by cewing** [comment]
@cguardia, @pwilkins  I do not know that we can yet.  If we do, we'll need to figure out how to fork the display logic for a given "course" based on whether it's a CCX or a regular course.  CCXs do not support the same exact API that a course does (yet).  

I think what's likely happening here is that there is a single enrollment only for a CCX.  That enrollment is being found once by the list of CCXs that the student is enrolled in (that's our code) and that is producing the second listing in the screenshot.  The *same* enrollment is also being found by the code that lists a student's enrollments in courses, and it is returning the course from which the ccx was made, which results in the first listing.  

If we are going to get rid of our list of CCX enrollments on the dashboard, then we'll need to figure out how to make enrollments for CCXs that appear in that first list grab their display information from the CCX, not from the course object.

**Comment by pdpinch** [comment]
That seems like a likely explanation, as you see in the screenshot there's different titles and start dates, but the same links.

**Comment by jetej** [comment]
Enrolling a student with the "Student List Management" works but "revoking access" does not.  At the moment, it is only possible to unenroll a student using "Batch Enrollment".  In addition, after the student is unenrolled, the student still appears in "Student List Management" despite having no access to the CCX.

**Comment by cguardia** [comment]
For the dashboard, am I correct that students are never supposed to join both the regular course and the CCX at the same time? If that's the case, should we look into modifying the main dashboard to show the CCX info instead when the user is actually enrolled in that?

**Comment by cewing** [comment]
I'm not sure that's a safe assumption, though obviously @pdpinch would know better.  

I **do** think that if we've retrieved the enrollment object using a CCXLocator rather than a CourseID object, we should be able to assert that we want to see the CCX data.

**Comment by pdpinch** [comment]
Assuming the student shouldn't enroll in the master course is an acceptable compromise, BUT I would prefer a solution that allows the student to enroll in the CCX without an enrollment in the master course. Is that feasible? 

In other words, I can live with hiding the incidental enrollment, but I'd rather not require it at all. 

----------------
I am mobile.

**Comment by pdpinch** [comment]
In further testing, I've confirmed that my CCX student, when enrolled in the CCX is not enrolled in the master course. This is a good thing, but it makes the duplicate in the student dashboard all the more puzzling.

**Comment by cguardia** [comment]
In that case, I propose to remove our list of courses form the dashboard, then change the original dashboard to check for a ccx for each course and use its information if found. We can do this on the same branch that we are working on, and see how it goes.

**Comment by cewing** [comment]
@pdpinch I actually don't think it's completely surprising.  But it is troubling.  Here's what I believe is happening.

The list of 'courses' in which a student is enrolled is arrived at by selecting records from the enrollment table, and then looking up the courses by 'course id'.  Because the CCXLocator (which is the custom version of a course id for a ccx) is "unwrapped" on the way through the modulestore wrapper the information for the course from which a ccx was created is found when looking up the enrollment that has a CCX id.  The student isn't actually enrolled in that course, but the information for that course is retrieved and displayed.  The links are rendered correctly for the CCX because the CCXLocator is re-wrapped as the data for the course comes out of the modulestore wrapper.

I agree with @cguardia that the best approach is to make the dashboard listing of courses aware of the difference between a CCXLocator and a regular Course ID, and then to remove the secondary listing of CCXs.

**Comment by pdpinch** [comment]
Ok. This makes sense to me now, and I agree with the solution. 

I presume the conditional in the dashboard listing template will be keyed off the use of a CCXLocator instead of a CourseLocator? When there is a CCXLocator, we want to be sure to display the overridden course name and start dates (as we do currently in the "secondary listing of CCXs").



> I agree with @cguardia <https://github.com/cguardia> that the best approach is to make the dashboard listing of courses aware of the difference between a CCXLocator and a regular Course ID, and then to remove the secondary listing of CCXs.

**Comment by cguardia** [comment]
My plan is, for each course, we will check if the course_id is a CCXLocator. If it is, we will use the CCX data, if not we'll simply use the old display code to show the course info.

**Comment by cewing** [comment]
This is sensible.  There is an alternate template used by our listing that should be used an example of how to display the CCX information.  It skips a great deal of the display options for the standard course because those options don't make sense for a CCX

**Comment by cguardia** [comment]
I did the dashboard change and pushed into this same remove-ccx-enrollment branch. It seems to work OK.

**Comment by pdpinch** [comment]
Great! I think this is ready to be PRed to edX. 

----------------
I am mobile.

**Comment by cguardia** [comment]
@pdpinch PR was just made. It took a bit longer than expected because it turns out edx very recently refactored the dashboard code and I had to make sure what we did worked under the new structure.

**Comment by cguardia** [comment]
https://github.com/edx/edx-platform/pull/8986

**Comment by pwilkins** [comment]
I'm encountering the same issue that @jetej reported above.  Revoking access in CCX doesn't remove the student from the ``Student List Management`` control even though the student can no longer view the CCX in their dashboard.

**Comment by cguardia** [comment]
@pdpinch I added a data migration, rebased and have all tests passing. Should I continue fixing issues like the one above as part of this PR?

**Comment by pdpinch** [comment]
We opened a new issue for this (#119) and it doesn't seem to me to be a blocker, so I'd like to handle it in another PR.

**Comment by pdpinch** [comment]
Closed via 34526ddf6a0be6585715c058025b86ea93e8a57e
